### PR TITLE
feat(frontend): use trait to impl predicate pushdown.

### LINKED ISSUE #2502: RFC: make predicates pushdown a trait and impl it for PlanNode.
We currently using rules to push predicates down. One limitation of this method is that PlanNodes must satisfy some patterns(e.g. Filter->Join) so that predicates can be pushed down. This method can solve all of the cases but we have to match too many patterns, which is lack of flexbility and can be a little annoying. Besides, it's easy to miss some patterns. A related issue is #2463.

To eliminate this limitation, We can have a trait named `PredicatesPushdown` which has a function named `push_down`, and implement it for every PlanNode, just like `prune_col`. The signature of `push_down` can be like this: `push_down(&self, predicates) -> PlanRef`. 

There are three kinds of predicates: 
1. those can't be pushed down. We just create a LogicalFilter for them above the current PlanNode.
2. those can be merged with current PlanNode(e.g. LogicalJoin). We just merge the predicates with the Condition of it.
3. those can be pushed down. We pass them to current PlanNode's child.

This method is expected to solve #2463, and gives more flexiblity which is useful if we want to support more rules to push predicates down.

### PR DESCRIPTION
**Pull request by likg227**
## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
As what #2502 said, we can use trait to impl predicate pushdown. 
This pr only migrate original rules related to predicate pushdown, and we can use the trait to push predicate through LogicalLimit and merge filter with scan etc.
Because the main functionality of this pr is migrate rules, I did't add many tests except for FilterJoin.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
close #2502

### CODE REVIEW

**Comment by likg227** [comment]
It seems that I can't directly delete FilterProject and FilterJoin because of LogicalMutiJoin...... cc @jon-chuang

**Comment by skyzh** [comment]
I think filter push-down is expected to be done after reordering.

**Comment by st1page** [comment]
> I think filter push-down is expected to be done after reordering.

+1

**Review by st1page — APPROVED** [review]
LGTM

File: src/frontend/src/optimizer/plan_node/predicate_pushdown.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    pushed_predicate: Condition,
+) -> PlanRef {
+    let new_input = node.input().predicate_pushdown(pushed_predicate);
+    let new_node = node.clone_with_input(new_input);
+    LogicalFilter::create(Rc::new(new_node), filter_predicate)
+}
+
+pub fn gen_filter_above(node: PlanRef, predicate: Condition) -> PlanRef {
+    LogicalFilter::create(node, predicate)
+}
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/predicate_pushdown.rs:56]: The helper functions seem not very useful，especially the second one.. I suggest to in-line them or add some comments

File: src/frontend/src/optimizer/plan_node/predicate_pushdown.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+use super::*;
+use crate::utils::Condition;
+use crate::{for_batch_plan_nodes, for_stream_plan_nodes};
+/// The trait for predicate pushdown, only logical plan node will use it, though all plan node impl
+/// it.
+pub trait PredicatePushdown {
+    /// Push predicate down for every logical plan node.
+    ///
+    /// If you don't know how to implement this trait, you can just use [`gen_new_node`] or
+    /// [`gen_filter_above`] depending on whether the plan node has ONLY one input or not.
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/predicate_pushdown.rs:26]: Please describe the cases more accurately (maybe copy part of the descriptions in RFC), and remove "If you don't know how to implement". Use precise and concise words.

**Review by xxchan — CHANGES_REQUESTED** [review]

File: src/frontend/src/optimizer/plan_node/predicate_pushdown.rs (hunk 1 above)
Comment (likg227) [reply src/frontend/src/optimizer/plan_node/predicate_pushdown.rs:56]: I will inline them.

File: src/frontend/src/optimizer/plan_node/predicate_pushdown.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    filter_predicate: Condition,
+    pushed_predicate: Condition,
+) -> PlanRef {
+    let new_input = node.input().predicate_pushdown(pushed_predicate);
+    let new_node = node.clone_with_input(new_input);
+    LogicalFilter::create(Rc::new(new_node), filter_predicate)
+}
+
+#[inline]
+pub fn gen_filter(node: PlanRef, predicate: Condition) -> PlanRef {
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/predicate_pushdown.rs:63]: I want you to inline it instead of the compiler 😄😆

Comment (xxchan) [reply src/frontend/src/optimizer/plan_node/predicate_pushdown.rs:63]: Why not create it directly

Comment (likg227) [reply src/frontend/src/optimizer/plan_node/predicate_pushdown.rs:63]: I just want abstract them as helper functions, which serve as default implementation........

Comment (xxchan) [reply src/frontend/src/optimizer/plan_node/predicate_pushdown.rs:63]: I think it's unnecessary. Just be straightforward.

Comment (likg227) [reply src/frontend/src/optimizer/plan_node/predicate_pushdown.rs:63]: Ok. I will change it.

File: src/frontend/src/optimizer/plan_node/predicate_pushdown.rs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    ///
+    /// Sometimes, you will find it hard or just want to delay the implementation of predicate
+    /// pushdown, then you can use [`gen_filter_and_pushdown`] to generate a `LogicalFilter` above
+    /// current node and do predicate pushdown for its input.
+    ///
+    /// Also, if current logcial plan node doesn't have any input(e.g. `LogicalScan`), you can use
+    /// [`gen_filter`] to generate a `LogicalFilter` above it.
+    ///
+    /// Please note that `LogicalFilter::create` will NOT create a `LogicalFilter` if `predicate`
+    /// is always true, so feel free to use these two helper functions.
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/predicate_pushdown.rs:33]: Still feel verbose. There's no need to leave such pedagogical guides. Just describe behavior. e.g. the three different kinds of operators as you mentioned in RFC

**Review by xxchan — APPROVED** [review]

**Review by xiangjinwu — APPROVED** [review]
Rest lgtm. We can do the cleanup in a followup.

**Pull request merged**